File: Legacy/Source/SocketManager.vb
Option Strict On
Option Explicit On

Imports System.Net
Imports System.Net.Sockets
Imports System.Text
Imports System.Threading
Imports System.Collections.Generic
Imports System.Timers

''' <summary>
''' Manages socket connections for the server.
''' </summary>
Public Class SocketManager
    ' Simple message structure for the queue
    Public Class QueuedMessage
        Public SocketID As Integer
        Public Data() As Byte

        Public Sub New(socketId As Integer, data() As Byte)
            Me.SocketID = socketId
            Me.Data = data
        End Sub
    End Class

    ' State object for reading client data asynchronously
    Public Class SocketState
        ' Client socket.
        Public Socket As Socket = Nothing
        ' Size of receive buffer.
        Public Const BufferSize As Integer = 8192
        ' Receive buffer.
        Public Buffer(BufferSize) As Byte
        ' ID associated with this socket
        Public ID As Integer = - 1
        ' Flag to track if socket is being closed
        Public IsClosing As Boolean = False

        Public Sub New(socket As Socket, id As Integer)
            Me.Socket = socket
            Me.ID = id
        End Sub
    End Class

    Private Shared _nextID As Integer = 1
    Private Shared _listener As Socket = Nothing
    Private Shared _listeningPort As Integer = 0

    ' Maps socket IDs to socket state objects
    Public Shared SocketStates As New Dictionary(Of Integer, SocketState)

    ' Pending operations tracking
    Private Shared _pendingReceives As New HashSet(Of Integer)
    Private Shared _pendingSends As New HashSet(Of Integer)

    ' Simple synchronized message queue
    Private Shared _messageQueue As New Queue(Of QueuedMessage)
    Private Shared _messageQueueLock As New Object()
    Private Shared _messageTimer As System.Timers.Timer

    ' Lock objects
    Private Shared _socketStatesLock As New Object()
    Private Shared _pendingReceivesLock As New Object()
    Private Shared _pendingSendsLock As New Object()

    ' Events
    Public Shared Event ConnectionReceived(socketID As Integer, clientIP As String)
    Public Shared Event DataReceived(socketID As Integer, data() As Byte)
    Public Shared Event ConnectionClosed(socketID As Integer)
    Public Shared Event ServerError(errorMessage As String)

    ' Flag to control if message processing is enabled
    Public Shared ProcessMessagesEnabled As Boolean = True

    ''' <summary>
    ''' Initializes the socket server to listen on the specified port.
    ''' </summary>
    Public Shared Sub Initialize(port As Integer)
        Encoding.RegisterProvider(CodePagesEncodingProvider.Instance)
        _listeningPort = port

        ' Initialize the message queue timer using System.Timers.Timer
        _messageTimer = New System.Timers.Timer(10) ' Check queue every 10ms
        AddHandler _messageTimer.Elapsed, AddressOf ProcessMessageQueue
        _messageTimer.AutoReset = True
        _messageTimer.Start()

        StartListening()
    End Sub

    ''' <summary>
    ''' Process queued messages on the main thread
    ''' </summary>
    Private Shared Sub ProcessMessageQueue(sender As Object, e As ElapsedEventArgs)
        ' Skip if processing is disabled
        If Not ProcessMessagesEnabled Then
            Return
        End If

        ' Process all available messages in one go
        While True
            Dim message As QueuedMessage = Nothing

            SyncLock _messageQueueLock
                If _messageQueue.Count > 0 Then
                    message = _messageQueue.Dequeue()
                Else
                    Exit While
                End If
            End SyncLock

            ' Process message on the main thread
            If message IsNot Nothing Then
                RaiseEvent DataReceived(message.SocketID, message.Data)
            End If
        End While
    End Sub

    ''' <summary>
    ''' Starts listening for incoming connections.
    ''' </summary>
    Public Shared Sub StartListening()
        Try
            ' Create an endpoint for the server
            Dim localEndPoint As New IPEndPoint(IPAddress.Any, _listeningPort)

            ' Create a TCP/IP socket
            _listener = New Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp)

            ' Bind the socket to the local endpoint and listen for incoming connections
            _listener.Bind(localEndPoint)
            _listener.Listen(100)

            ' Start an asynchronous socket to listen for connections
            _listener.BeginAccept(New AsyncCallback(AddressOf AcceptCallback), _listener)
        Catch ex As Exception
            RaiseEvent ServerError("Error starting server: " & ex.Message)
        End Try
    End Sub

    ''' <summary>
    ''' Callback method executed when a connection request is received.
    ''' </summary>
    Private Shared Sub AcceptCallback(ar As IAsyncResult)
        Try
            ' Get the socket that handles the client request
            Dim listener As Socket = CType(ar.AsyncState, Socket)

            ' Check if the listener is still valid
            If listener Is Nothing OrElse Not listener.IsBound Then
                Return
            End If

            Dim socket As Socket = listener.EndAccept(ar)

            ' Generate a unique ID for this socket
            Dim socketID As Integer = _nextID
            Interlocked.Increment(_nextID)

            ' Create the state object for this client
            Dim state As New SocketState(socket, socketID)

            ' Store the socket state in the dictionary
            SyncLock _socketStatesLock
                SocketStates(socketID) = state
            End SyncLock

            ' Get client IP
            Dim clientIP As String = CType(socket.RemoteEndPoint, IPEndPoint).Address.ToString()

            ' Notify about the new connection
            RaiseEvent ConnectionReceived(socketID, clientIP)

            ' Begin receiving data from the client
            StartReceive(state)

            ' Continue listening for connections
            If _listener IsNot Nothing AndAlso _listener.IsBound Then
                _listener.BeginAccept(New AsyncCallback(AddressOf AcceptCallback), _listener)
            End If
        Catch ex As ObjectDisposedException
            ' Socket was closed
        Catch ex As Exception
            RaiseEvent ServerError("Error accepting connection: " & ex.Message)

            ' Try to continue listening
            Try
                If _listener IsNot Nothing AndAlso _listener.IsBound Then
                    _listener.BeginAccept(New AsyncCallback(AddressOf AcceptCallback), _listener)
                End If
            Catch
                ' Ignore if we can't continue
            End Try
        End Try
    End Sub

    ''' <summary>
    ''' Starts an asynchronous receive operation on a socket.
    ''' </summary>
    Private Shared Sub StartReceive(state As SocketState)
        ' Mark this socket as having a pending receive
        Dim socketID As Integer = state.ID

        SyncLock _pendingReceivesLock
            If _pendingReceives.Contains(socketID) Then
                ' Already receiving on this socket, don't start another receive
                Return
            End If

            _pendingReceives.Add(socketID)
        End SyncLock

        Try
            If state.Socket Is Nothing OrElse Not state.Socket.Connected OrElse state.IsClosing Then
                ' Socket is no longer valid or is being closed, don't start a receive
                SyncLock _pendingReceivesLock
                    _pendingReceives.Remove(socketID)
                End SyncLock
                Return
            End If

            ' Begin receiving data from the client
            state.Socket.BeginReceive(state.Buffer, 0, SocketState.BufferSize, 0,
                                      New AsyncCallback(AddressOf ReadCallback), state)
        Catch ex As ObjectDisposedException
            ' Socket was closed
            SyncLock _pendingReceivesLock
                _pendingReceives.Remove(socketID)
            End SyncLock
            CloseSocketInternal(socketID)
        Catch ex As Exception
            ' Error during BeginReceive
            SyncLock _pendingReceivesLock
                _pendingReceives.Remove(socketID)
            End SyncLock
            RaiseEvent ServerError("Error starting receive: " & ex.Message)
            CloseSocketInternal(socketID)
        End Try
    End Sub

    ''' <summary>
    ''' Callback method executed when data is read from the client.
    ''' </summary>
    Private Shared Sub ReadCallback(ar As IAsyncResult)
        If ar Is Nothing Then Return

        Dim state As SocketState = Nothing
        Dim socketID As Integer = - 1

        Try
            state = CType(ar.AsyncState, SocketState)
            If state Is Nothing Then Return

            socketID = state.ID

            ' Remove from pending receives
            SyncLock _pendingReceivesLock
                _pendingReceives.Remove(socketID)
            End SyncLock

            ' Check if the socket is valid
            If state.Socket Is Nothing OrElse Not state.Socket.Connected OrElse state.IsClosing Then
                CloseSocketInternal(socketID)
                Return
            End If

            ' Read data from the client socket
            Dim bytesRead As Integer = state.Socket.EndReceive(ar)

            If bytesRead > 0 Then
                ' Create a copy of the data received
                Dim data(bytesRead - 1) As Byte
                Array.Copy(state.Buffer, data, bytesRead)

                ' Add to message queue instead of raising event directly
                EnqueueMessage(socketID, data)

                ' Start receiving more data
                StartReceive(state)
            Else
                ' Connection closed by the client
                CloseSocketInternal(socketID)
            End If
        Catch ex As ObjectDisposedException
            ' Socket was closed
            If socketID >= 0 Then CloseSocketInternal(socketID)
        Catch ex As SocketException
            ' Socket error
            If socketID >= 0 Then CloseSocketInternal(socketID)
        Catch ex As Exception
            ' Other error
            RaiseEvent ServerError("Error reading from socket: " & ex.Message)
            If socketID >= 0 Then CloseSocketInternal(socketID)
        End Try
    End Sub

    ''' <summary>
    ''' Adds a message to the queue for processing on the main thread
    ''' </summary>
    Private Shared Sub EnqueueMessage(socketID As Integer, data() As Byte)
        SyncLock _messageQueueLock
            _messageQueue.Enqueue(New QueuedMessage(socketID, data))
        End SyncLock
    End Sub

    ''' <summary>
    ''' Manually process any pending messages
    ''' </summary>
    Public Shared Sub ProcessPendingMessages()
        ' Only call this from the main thread
        ProcessMessageQueue(Nothing, Nothing)
    End Sub

    ''' <summary>
    ''' Sends data to a client.
    ''' </summary>
    Public Shared Function SendData(socketID As Integer, data() As Byte) As Boolean
        If data Is Nothing OrElse data.Length = 0 Then
            Return True ' Nothing to send, consider it a success
        End If

        Dim state As SocketState = Nothing

        SyncLock _socketStatesLock
            If Not SocketStates.TryGetValue(socketID, state) Then
                Return False
            End If

            ' Don't send data to a socket that's being closed
            If state.IsClosing Then
                Return False
            End If
        End SyncLock

        ' Mark this socket as having a pending send
        SyncLock _pendingSendsLock
            _pendingSends.Add(socketID)
        End SyncLock

        Try
            If state.Socket Is Nothing OrElse Not state.Socket.Connected Then
                SyncLock _pendingSendsLock
                    _pendingSends.Remove(socketID)
                End SyncLock
                CloseSocketInternal(socketID)
                Return False
            End If

            state.Socket.BeginSend(data, 0, data.Length, 0, New AsyncCallback(AddressOf SendCallback), state)
            Return True
        Catch ex As ObjectDisposedException
            ' Socket was closed
            SyncLock _pendingSendsLock
                _pendingSends.Remove(socketID)
            End SyncLock
            CloseSocketInternal(socketID)
            Return False
        Catch ex As Exception
            RaiseEvent ServerError("Error sending data: " & ex.Message)
            SyncLock _pendingSendsLock
                _pendingSends.Remove(socketID)
            End SyncLock
            CloseSocketInternal(socketID)
            Return False
        End Try
    End Function

    ''' <summary>
    ''' Callback method executed when data send operation completes.
    ''' </summary>
    Private Shared Sub SendCallback(ar As IAsyncResult)
        If ar Is Nothing Then Return

        Dim state As SocketState = Nothing
        Dim socketID As Integer = - 1

        Try
            state = CType(ar.AsyncState, SocketState)
            If state Is Nothing Then Return

            socketID = state.ID

            ' Remove from pending sends
            SyncLock _pendingSendsLock
                _pendingSends.Remove(socketID)
            End SyncLock

            ' Check if the socket is still valid
            If state.Socket Is Nothing OrElse state.IsClosing Then
                Return
            End If

            ' Complete sending the data to the remote device
            state.Socket.EndSend(ar)
        Catch ex As ObjectDisposedException
            ' Socket was closed
        Catch ex As SocketException
            ' Socket error
            If socketID >= 0 Then CloseSocketInternal(socketID)
        Catch ex As Exception
            RaiseEvent ServerError("Error completing send: " & ex.Message)
            If socketID >= 0 Then CloseSocketInternal(socketID)
        End Try
    End Sub

    ''' <summary>
    ''' Closes a client socket.
    ''' </summary>
    Public Shared Sub CloseSocket(socketID As Integer)
        CloseSocketInternal(socketID)
    End Sub

    ''' <summary>
    ''' Internal implementation of closing a socket.
    ''' </summary>
    Private Shared Sub CloseSocketInternal(socketID As Integer)
        Dim state As SocketState = Nothing

        SyncLock _socketStatesLock
            If Not SocketStates.TryGetValue(socketID, state) Then
                Return
            End If

            ' Mark the socket as closing to prevent further operations
            state.IsClosing = True
        End SyncLock

        ' Check if there are pending operations on this socket
        Dim hasPendingOperations As Boolean = False

        SyncLock _pendingReceivesLock
            hasPendingOperations = hasPendingOperations OrElse _pendingReceives.Contains(socketID)
        End SyncLock

        SyncLock _pendingSendsLock
            hasPendingOperations = hasPendingOperations OrElse _pendingSends.Contains(socketID)
        End SyncLock

        ' If we have pending operations, we'll wait for them to complete
        If hasPendingOperations Then
            ' Schedule a check to see if pending operations have finished
            Dim timer As New System.Timers.Timer(100) ' Check again after 100ms
            AddHandler timer.Elapsed, Sub(sender, e)
                timer.Stop()
                timer.Dispose()
                FinishCloseSocket(socketID)
            End Sub
            timer.AutoReset = False
            timer.Start()
        Else
            ' No pending operations, proceed with closing
            FinishCloseSocket(socketID)
        End If
    End Sub

    ''' <summary>
    ''' Completes the socket closing process after any pending operations are done.
    ''' </summary>
    Private Shared Sub FinishCloseSocket(socketID As Integer)
        Dim state As SocketState = Nothing

        SyncLock _socketStatesLock
            If Not SocketStates.TryGetValue(socketID, state) Then
                Return
            End If

            SocketStates.Remove(socketID)
        End SyncLock

        ' Check if there are still pending operations
        Dim hasPendingOperations As Boolean = False

        SyncLock _pendingReceivesLock
            hasPendingOperations = hasPendingOperations OrElse _pendingReceives.Contains(socketID)
        End SyncLock

        SyncLock _pendingSendsLock
            hasPendingOperations = hasPendingOperations OrElse _pendingSends.Contains(socketID)
        End SyncLock

        ' If we still have pending operations, wait a bit longer
        If hasPendingOperations Then
            Dim timer As New System.Timers.Timer(100) ' Check again after 100ms
            AddHandler timer.Elapsed, Sub(sender, e)
                timer.Stop()
                timer.Dispose()
                FinishCloseSocket(socketID)
            End Sub
            timer.AutoReset = False
            timer.Start()
            Return
        End If

        ' Actually close the socket
        Try
            If state IsNot Nothing AndAlso state.Socket IsNot Nothing Then
                state.Socket.Shutdown(SocketShutdown.Both)
                state.Socket.Close()
                state.Socket = Nothing
            End If
        Catch ex As ObjectDisposedException
            ' Socket was already disposed
        Catch ex As SocketException
            ' Socket error during shutdown/close
        Catch ex As Exception
            ' Other errors during shutdown/close
        End Try

        ' Notify about the closed connection
        RaiseEvent ConnectionClosed(socketID)
    End Sub

    ''' <summary>
    ''' Converts a string to a byte array using Windows-1252 encoding.
    ''' </summary>
    Public Shared Function StringToBytes(text As String) As Byte()
        Return Encoding.GetEncoding("Windows-1252").GetBytes(text)
    End Function

    ''' <summary>
    ''' Converts a byte array to a string using Windows-1252 encoding.
    ''' </summary>
    Public Shared Function BytesToString(bytes() As Byte) As String
        Return Encoding.GetEncoding("Windows-1252").GetString(bytes)
    End Function

    ''' <summary>
    ''' Shutdown the socket server.
    ''' </summary>
    Public Shared Sub Shutdown()
        ' Stop the message processing timer
        If _messageTimer IsNot Nothing Then
            _messageTimer.Stop()
            RemoveHandler _messageTimer.Elapsed, AddressOf ProcessMessageQueue
            _messageTimer.Dispose()
            _messageTimer = Nothing
        End If

        ' Clear the message queue
        SyncLock _messageQueueLock
            _messageQueue.Clear()
        End SyncLock

        ' Close all client sockets
        Dim allSocketIDs As New List(Of Integer)

        SyncLock _socketStatesLock
            Dim kvp As KeyValuePair(Of Integer, SocketState)
            For Each kvp In SocketStates
                allSocketIDs.Add(kvp.Key)
            Next
        End SyncLock

        Dim socketID As Integer
        For Each socketID In allSocketIDs
            CloseSocketInternal(socketID)
        Next

        ' Close the listening socket
        Try
            If _listener IsNot Nothing Then
                _listener.Close()
                _listener = Nothing
            End If
        Catch
            ' Ignore errors during shutdown
        End Try
    End Sub
End Class

File: Legacy/Source/SocketManager.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading;
using System.Timers;
using Timer = System.Timers.Timer;

namespace Legacy;

/// <summary>
///     Manages socket connections for the server.
/// </summary>
public class SocketManager
{
    public delegate void ConnectionClosedEventHandler(int socketID);

    public delegate void ConnectionReceivedEventHandler(int socketID, string clientIP);

    public delegate void DataReceivedEventHandler(int socketID, byte[] data);

    public delegate void ServerErrorEventHandler(string errorMessage);

    private static int _nextID = 1;
    private static Socket _listener;
    private static int _listeningPort;

    // Maps socket IDs to socket state objects
    public static Dictionary<int, SocketState> SocketStates = new();

    // Pending operations tracking
    private static readonly HashSet<int> _pendingReceives = new();
    private static readonly HashSet<int> _pendingSends = new();

    // Simple synchronized message queue
    private static readonly Queue<QueuedMessage> _messageQueue = new();
    private static readonly object _messageQueueLock = new();
    private static Timer _messageTimer;

    // Lock objects
    private static readonly object _socketStatesLock = new();
    private static readonly object _pendingReceivesLock = new();
    private static readonly object _pendingSendsLock = new();

    // Flag to control if message processing is enabled
    public static bool ProcessMessagesEnabled = true;

    // Events
    public static event ConnectionReceivedEventHandler ConnectionReceived;
    public static event DataReceivedEventHandler DataReceived;
    public static event ConnectionClosedEventHandler ConnectionClosed;
    public static event ServerErrorEventHandler ServerError;

    /// <summary>
    ///     Initializes the socket server to listen on the specified port.
    /// </summary>
    public static void Initialize(int port)
    {
        Encoding.RegisterProvider(CodePagesEncodingProvider.Instance);
        _listeningPort = port;

        // Initialize the message queue timer using System.Timers.Timer
        _messageTimer = new Timer(10d); // Check queue every 10ms
        _messageTimer.Elapsed += ProcessMessageQueue;
        _messageTimer.AutoReset = true;
        _messageTimer.Start();

        StartListening();
    }

    /// <summary>
    ///     Process queued messages on the main thread
    /// </summary>
    private static void ProcessMessageQueue(object sender, ElapsedEventArgs e)
    {
        // Skip if processing is disabled
        if (!ProcessMessagesEnabled) return;

        // Process all available messages in one go
        while (true)
        {
            QueuedMessage message = null;

            lock (_messageQueueLock)
            {
                if (_messageQueue.Count > 0)
                    message = _messageQueue.Dequeue();
                else
                    break;
            }

            // Process message on the main thread
            if (message is not null) DataReceived?.Invoke(message.SocketID, message.Data);
        }
    }

    /// <summary>
    ///     Starts listening for incoming connections.
    /// </summary>
    public static void StartListening()
    {
        try
        {
            // Create an endpoint for the server
            var localEndPoint = new IPEndPoint(IPAddress.Any, _listeningPort);

            // Create a TCP/IP socket
            _listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);

            // Bind the socket to the local endpoint and listen for incoming connections
            _listener.Bind(localEndPoint);
            _listener.Listen(100);

            // Start an asynchronous socket to listen for connections
            _listener.BeginAccept(AcceptCallback, _listener);
        }
        catch (Exception ex)
        {
            ServerError?.Invoke("Error starting server: " + ex.Message);
        }
    }

    /// <summary>
    ///     Callback method executed when a connection request is received.
    /// </summary>
    private static void AcceptCallback(IAsyncResult ar)
    {
        try
        {
            // Get the socket that handles the client request
            var listener = (Socket)ar.AsyncState;

            // Check if the listener is still valid
            if (listener is null || !listener.IsBound) return;

            var socket = listener.EndAccept(ar);

            // Generate a unique ID for this socket
            var socketID = _nextID;
            Interlocked.Increment(ref _nextID);

            // Create the state object for this client
            var state = new SocketState(socket, socketID);

            // Store the socket state in the dictionary
            lock (_socketStatesLock)
            {
                SocketStates[socketID] = state;
            }

            // Get client IP
            var clientIP = ((IPEndPoint)socket.RemoteEndPoint).Address.ToString();

            // Notify about the new connection
            ConnectionReceived?.Invoke(socketID, clientIP);

            // Begin receiving data from the client
            StartReceive(state);

            // Continue listening for connections
            if (_listener is not null && _listener.IsBound) _listener.BeginAccept(AcceptCallback, _listener);
        }
        catch (ObjectDisposedException ex)
        {
        }
        // Socket was closed
        catch (Exception ex)
        {
            ServerError?.Invoke("Error accepting connection: " + ex.Message);

            // Try to continue listening
            try
            {
                if (_listener is not null && _listener.IsBound) _listener.BeginAccept(AcceptCallback, _listener);
            }
            catch
            {
                // Ignore if we can't continue
            }
        }
    }

    /// <summary>
    ///     Starts an asynchronous receive operation on a socket.
    /// </summary>
    private static void StartReceive(SocketState state)
    {
        // Mark this socket as having a pending receive
        var socketID = state.ID;

        lock (_pendingReceivesLock)
        {
            if (_pendingReceives.Contains(socketID))
                // Already receiving on this socket, don't start another receive
                return;

            _pendingReceives.Add(socketID);
        }

        try
        {
            if (state.Socket is null || !state.Socket.Connected || state.IsClosing)
            {
                // Socket is no longer valid or is being closed, don't start a receive
                lock (_pendingReceivesLock)
                {
                    _pendingReceives.Remove(socketID);
                }

                return;
            }

            // Begin receiving data from the client
            state.Socket.BeginReceive(state.Buffer, 0, SocketState.BufferSize, 0, ReadCallback, state);
        }
        catch (ObjectDisposedException ex)
        {
            // Socket was closed
            lock (_pendingReceivesLock)
            {
                _pendingReceives.Remove(socketID);
            }

            CloseSocketInternal(socketID);
        }
        catch (Exception ex)
        {
            // Error during BeginReceive
            lock (_pendingReceivesLock)
            {
                _pendingReceives.Remove(socketID);
            }

            ServerError?.Invoke("Error starting receive: " + ex.Message);
            CloseSocketInternal(socketID);
        }
    }

    /// <summary>
    ///     Callback method executed when data is read from the client.
    /// </summary>
    private static void ReadCallback(IAsyncResult ar)
    {
        if (ar is null)
            return;

        SocketState state = null;
        var socketID = -1;

        try
        {
            state = (SocketState)ar.AsyncState;
            if (state is null)
                return;

            socketID = state.ID;

            // Remove from pending receives
            lock (_pendingReceivesLock)
            {
                _pendingReceives.Remove(socketID);
            }

            // Check if the socket is valid
            if (state.Socket is null || !state.Socket.Connected || state.IsClosing)
            {
                CloseSocketInternal(socketID);
                return;
            }

            // Read data from the client socket
            var bytesRead = state.Socket.EndReceive(ar);

            if (bytesRead > 0)
            {
                // Create a copy of the data received
                var data = new byte[bytesRead];
                Array.Copy(state.Buffer, data, bytesRead);

                // Add to message queue instead of raising event directly
                EnqueueMessage(socketID, data);

                // Start receiving more data
                StartReceive(state);
            }
            else
            {
                // Connection closed by the client
                CloseSocketInternal(socketID);
            }
        }
        catch (ObjectDisposedException ex)
        {
            // Socket was closed
            if (socketID >= 0)
                CloseSocketInternal(socketID);
        }
        catch (SocketException ex)
        {
            // Socket error
            if (socketID >= 0)
                CloseSocketInternal(socketID);
        }
        catch (Exception ex)
        {
            // Other error
            ServerError?.Invoke("Error reading from socket: " + ex.Message);
            if (socketID >= 0)
                CloseSocketInternal(socketID);
        }
    }

    /// <summary>
    ///     Adds a message to the queue for processing on the main thread
    /// </summary>
    private static void EnqueueMessage(int socketID, byte[] data)
    {
        lock (_messageQueueLock)
        {
            _messageQueue.Enqueue(new QueuedMessage(socketID, data));
        }
    }

    /// <summary>
    ///     Manually process any pending messages
    /// </summary>
    public static void ProcessPendingMessages()
    {
        // Only call this from the main thread
        ProcessMessageQueue(null, null);
    }

    /// <summary>
    ///     Sends data to a client.
    /// </summary>
    public static bool SendData(int socketID, byte[] data)
    {
        if (data is null || data.Length == 0) return true; // Nothing to send, consider it a success

        SocketState state = null;

        lock (_socketStatesLock)
        {
            if (!SocketStates.TryGetValue(socketID, out state)) return false;

            // Don't send data to a socket that's being closed
            if (state.IsClosing) return false;
        }

        // Mark this socket as having a pending send
        lock (_pendingSendsLock)
        {
            _pendingSends.Add(socketID);
        }

        try
        {
            if (state.Socket is null || !state.Socket.Connected)
            {
                lock (_pendingSendsLock)
                {
                    _pendingSends.Remove(socketID);
                }

                CloseSocketInternal(socketID);
                return false;
            }

            state.Socket.BeginSend(data, 0, data.Length, 0, SendCallback, state);
            return true;
        }
        catch (ObjectDisposedException ex)
        {
            // Socket was closed
            lock (_pendingSendsLock)
            {
                _pendingSends.Remove(socketID);
            }

            CloseSocketInternal(socketID);
            return false;
        }
        catch (Exception ex)
        {
            ServerError?.Invoke("Error sending data: " + ex.Message);
            lock (_pendingSendsLock)
            {
                _pendingSends.Remove(socketID);
            }

            CloseSocketInternal(socketID);
            return false;
        }
    }

    /// <summary>
    ///     Callback method executed when data send operation completes.
    /// </summary>
    private static void SendCallback(IAsyncResult ar)
    {
        if (ar is null)
            return;

        SocketState state = null;
        var socketID = -1;

        try
        {
            state = (SocketState)ar.AsyncState;
            if (state is null)
                return;

            socketID = state.ID;

            // Remove from pending sends
            lock (_pendingSendsLock)
            {
                _pendingSends.Remove(socketID);
            }

            // Check if the socket is still valid
            if (state.Socket is null || state.IsClosing) return;

            // Complete sending the data to the remote device
            state.Socket.EndSend(ar);
        }
        catch (ObjectDisposedException ex)
        {
        }
        // Socket was closed
        catch (SocketException ex)
        {
            // Socket error
            if (socketID >= 0)
                CloseSocketInternal(socketID);
        }
        catch (Exception ex)
        {
            ServerError?.Invoke("Error completing send: " + ex.Message);
            if (socketID >= 0)
                CloseSocketInternal(socketID);
        }
    }

    /// <summary>
    ///     Closes a client socket.
    /// </summary>
    public static void CloseSocket(int socketID)
    {
        CloseSocketInternal(socketID);
    }

    /// <summary>
    ///     Internal implementation of closing a socket.
    /// </summary>
    private static void CloseSocketInternal(int socketID)
    {
        SocketState state = null;

        lock (_socketStatesLock)
        {
            if (!SocketStates.TryGetValue(socketID, out state)) return;

            // Mark the socket as closing to prevent further operations
            state.IsClosing = true;
        }

        // Check if there are pending operations on this socket
        var hasPendingOperations = false;

        lock (_pendingReceivesLock)
        {
            hasPendingOperations = hasPendingOperations || _pendingReceives.Contains(socketID);
        }

        lock (_pendingSendsLock)
        {
            hasPendingOperations = hasPendingOperations || _pendingSends.Contains(socketID);
        }

        // If we have pending operations, we'll wait for them to complete
        if (hasPendingOperations)
        {
            // Schedule a check to see if pending operations have finished
            var timer = new Timer(100d); // Check again after 100ms
            timer.Elapsed += (sender, e) =>
            {
                timer.Stop();
                timer.Dispose();
                FinishCloseSocket(socketID);
            };
            timer.AutoReset = false;
            timer.Start();
        }
        else
        {
            // No pending operations, proceed with closing
            FinishCloseSocket(socketID);
        }
    }

    /// <summary>
    ///     Completes the socket closing process after any pending operations are done.
    /// </summary>
    private static void FinishCloseSocket(int socketID)
    {
        SocketState state = null;

        lock (_socketStatesLock)
        {
            if (!SocketStates.TryGetValue(socketID, out state)) return;

            SocketStates.Remove(socketID);
        }

        // Check if there are still pending operations
        var hasPendingOperations = false;

        lock (_pendingReceivesLock)
        {
            hasPendingOperations = hasPendingOperations || _pendingReceives.Contains(socketID);
        }

        lock (_pendingSendsLock)
        {
            hasPendingOperations = hasPendingOperations || _pendingSends.Contains(socketID);
        }

        // If we still have pending operations, wait a bit longer
        if (hasPendingOperations)
        {
            var timer = new Timer(100d); // Check again after 100ms
            timer.Elapsed += (sender, e) =>
            {
                timer.Stop();
                timer.Dispose();
                FinishCloseSocket(socketID);
            };
            timer.AutoReset = false;
            timer.Start();
            return;
        }

        // Actually close the socket
        try
        {
            if (state is not null && state.Socket is not null)
            {
                state.Socket.Shutdown(SocketShutdown.Both);
                state.Socket.Close();
                state.Socket = null;
            }
        }
        catch (ObjectDisposedException ex)
        {
        }
        // Socket was already disposed
        catch (SocketException ex)
        {
        }
        // Socket error during shutdown/close
        catch (Exception ex)
        {
            // Other errors during shutdown/close
        }

        // Notify about the closed connection
        ConnectionClosed?.Invoke(socketID);
    }

    /// <summary>
    ///     Converts a string to a byte array using Windows-1252 encoding.
    /// </summary>
    public static byte[] StringToBytes(string text)
    {
        return Encoding.GetEncoding("Windows-1252").GetBytes(text);
    }

    /// <summary>
    ///     Converts a byte array to a string using Windows-1252 encoding.
    /// </summary>
    public static string BytesToString(byte[] bytes)
    {
        return Encoding.GetEncoding("Windows-1252").GetString(bytes);
    }

    /// <summary>
    ///     Shutdown the socket server.
    /// </summary>
    public static void Shutdown()
    {
        // Stop the message processing timer
        if (_messageTimer is not null)
        {
            _messageTimer.Stop();
            _messageTimer.Elapsed -= ProcessMessageQueue;
            _messageTimer.Dispose();
            _messageTimer = null;
        }

        // Clear the message queue
        lock (_messageQueueLock)
        {
            _messageQueue.Clear();
        }

        // Close all client sockets
        var allSocketIDs = new List<int>();

        lock (_socketStatesLock)
        {
            foreach (var kvp in SocketStates)
                allSocketIDs.Add(kvp.Key);
        }

        foreach (var socketID in allSocketIDs)
            CloseSocketInternal(socketID);

        // Close the listening socket
        try
        {
            if (_listener is not null)
            {
                _listener.Close();
                _listener = null;
            }
        }
        catch
        {
            // Ignore errors during shutdown
        }
    }

    // Simple message structure for the queue
    public class QueuedMessage
    {
        public byte[] Data;
        public int SocketID;

        public QueuedMessage(int socketId, byte[] data)
        {
            SocketID = socketId;
            Data = data;
        }
    }

    // State object for reading client data asynchronously
    public class SocketState
    {
        // Size of receive buffer.
        public const int BufferSize = 8192;

        // Receive buffer.
        public byte[] Buffer = new byte[8193];

        // ID associated with this socket
        public int ID = -1;

        // Flag to track if socket is being closed
        public bool IsClosing;

        // Client socket.
        public Socket Socket;

        public SocketState(Socket socket, int id)
        {
            Socket = socket;
            ID = id;
        }
    }
}